File: source/cf/solutions/aws-mini-connected-factory-v1.0/lambdas/aws-mini-connected-factory-belt-serial-node/belt.js
const SerialPort = require('serialport');
const Readline = require('@serialport/parser-readline');

const events = require('events');

function getCharFor(speed, mode) {
    let char = '5';

    if (speed === 1) {
        if (mode === 1) {
            char = '4';
        } else if (mode === 2) {
            char = '5';
        } else if (mode === 3) {
            char = '6';
        }
    } else if (speed === 2) {
        if (mode === 1) {
            char = '3';
        } else if (mode === 2) {
            char = '5';
        } else if (mode === 3) {
            char = '7';
        }
    }

    return char;
}

class Belt extends events.EventEmitter {

    constructor(port = '/dev/null', baud = 115200, pollerFreq = 1000) {
        super();

        console.log('Belt: init');

        this.port = new SerialPort(port, {
            baudRate: baud
        });

        this.parser = this.port.pipe(new Readline({
            delimiter: '\n'
        }));

        this.SHADOW_DESIRED = {
            "mode": 2,
            "speed": 1
        };
        this.SHADOW_REPORTED = {
            "mode": 2,
            "speed": 1
        };

        this.parser.on('data', (data) => {
            this.parseIncomingSerialLine(data, (data) => {
                // console.log('writeSerial:', data);
                this.port.write(data);
            });
        });

        this.shadowPoller = setInterval(() => {
            this.port.write('1');
        }, pollerFreq);
    }

    parseIncomingSerialLine(data, serialWrite) {

        const TELEMETRY_STRING = ' [TelemetryTask] [BELT_TELEMETRY] ';
        const SHADOW_WATCHDOGTASK_STRING = ' [WatchdogTask] [BELT_SHADOW] ';
        const SHADOW_SERIALINPUTTASK_STRING = ' [SerialInputTask] [BELT_SHADOW] ';
        const tag = 'parseIncomingSerialLine:';

        try {
            // console.log(tag, data);
            this.emit('data', data);

            let beltData = null;

            if (data.includes(TELEMETRY_STRING + '{')) {
                beltData = JSON.parse(data.split(TELEMETRY_STRING)[1]);
            }
            if (data.includes(SHADOW_WATCHDOGTASK_STRING + '{')) {
                beltData = JSON.parse(data.split(SHADOW_WATCHDOGTASK_STRING)[1]);
            }
            if (data.includes(SHADOW_SERIALINPUTTASK_STRING + '{')) {
                beltData = JSON.parse(data.split(SHADOW_SERIALINPUTTASK_STRING)[1]);
            }

            if (beltData && beltData.hasOwnProperty('state') && beltData.state.hasOwnProperty('reported')) {

                const reported = beltData.state.reported;

                this.emit('shadow', {
                    type: 'info',
                    data: reported
                });

                let needToUpdateShadow = false;

                if (reported.hasOwnProperty('speed')) {
                    if (reported.speed !== 1 && reported.speed !== 2) {
                        // console.log(tag, 'Incorrect speed reported');
                        serialWrite(getCharFor(this.SHADOW_DESIRED.speed, this.SHADOW_DESIRED.mode));
                    } else {
                        if (this.SHADOW_REPORTED.speed != reported.speed) {
                            needToUpdateShadow = true;
                            this.SHADOW_REPORTED.speed = reported.speed;
                        }
                    }
                }
                if (reported.hasOwnProperty('mode')) {
                    if (reported.mode !== 1 && reported.mode !== 2 && reported.mode !== 3) {
                        // console.log(tag, 'Incorrect mode reported');
                        serialWrite(getCharFor(this.SHADOW_DESIRED.speed, this.SHADOW_DESIRED.mode));
                    } else {
                        if (this.SHADOW_REPORTED.mode != reported.mode) {
                            needToUpdateShadow = true;
                            this.SHADOW_REPORTED.mode = reported.mode;
                        }
                    }
                }
                if (needToUpdateShadow) {
                    this.emit('shadow', {
                        type: 'reported',
                        data: this.SHADOW_REPORTED
                    });
                    // ggIoT.updateThingShadow(
                    //     payload = {
                    //         'state': {
                    //             'reported': this.SHADOW_REPORTED
                    //         }
                    //     })
                }
            }

            if (beltData && beltData.hasOwnProperty('chassis')) {
                const chassis = beltData.chassis;

                this.emit('sensors', chassis);

                // if (chassis.hasOwnProperty('x') && chassis.hasOwnProperty('y') && chassis.hasOwnProperty('z')) {
                //    // ggIoT.publish(TOPIC_FOR_SENSORS, data);
                // }

            }
            // console.log(tag, 'desired vs. reported:', this.SHADOW_DESIRED, this.SHADOW_REPORTED);
            if (this.SHADOW_REPORTED.mode != this.SHADOW_DESIRED.mode || this.SHADOW_REPORTED.speed != this.SHADOW_DESIRED.speed) {
                serialWrite(getCharFor(this.SHADOW_DESIRED.speed, this.SHADOW_DESIRED.mode));
            }

        } catch (ex) {
            this.emit('error', ex);
            // console.log('parseIncomingSerialLine: ERROR:', ex);
            // ggIoT.exception(str(ex))
        }
    }

    parseIncomingShadow(data) {
        if (data.hasOwnProperty('state') && data.state.hasOwnProperty('desired')) {
            const desired = data.state.desired;

            if (desired.hasOwnProperty('mode') && desired.mode !== this.SHADOW_DESIRED.mode) {
                this.SHADOW_DESIRED.mode = desired.mode;
            }
            if (desired.hasOwnProperty('speed') && desired.speed !== this.SHADOW_DESIRED.speed) {
                this.SHADOW_DESIRED.speed = desired.speed;
            }

            this.emit('shadow', {
                type: 'desired',
                data: this.SHADOW_DESIRED
            });
            // ggIoT.info('New Shadow Received');
        }
    }

}

module.exports = Belt;

File: source/solutions/mini-connected-factory/lambdas/mini-connected-factory-belt-serial-python/belt.py
import json

TELEMETRY_STRING = " [BELT_TELEMETRY] "
SHADOW_STRING = " [BELT_SHADOW] "
PROXIMITY_STRING = " [BELT_PROXIMITY] "

class Belt:

    def __init__(self):
        self.SHADOW_DESIRED = {
            "mode": 2,
            "speed": 1
        }
        self.SHADOW_REPORTED = {
            "mode": 2,
            "speed": 1
        }
        self.SENSORS = {
            "speed": {
                "rpm": 0
            },
            "chassis": {
                "x": 0,
                "y": 0,
                "z": 0
            },
            "proximity": {
                "sensor1": 0,
                "sensor2": 0
            }
        }

    def getCharFor(self, speed, mode):
        char = "5"
        if speed == 1:
            if mode == 1:
                char = "4"
            elif mode == 2:
                char = "5"
            elif mode == 3:
                char = "6"
        elif speed == 2:
            if mode == 1:
                char = "3"
            elif mode == 2:
                char = "5"
            elif mode == 3:
                char = "7"

        return char

    def writeDesiredToSerial(self, serialWrite):
        char = self.getCharFor(speed=self.SHADOW_DESIRED["speed"], mode=self.SHADOW_DESIRED["mode"])
        print("Belt.parseSerial: Writing to serial: {}".format(char))
        serialWrite(char)

    def parseSerial(self, data, serialWrite, GGIOT, TOPIC_FOR_SENSORS):

        beltData = None
        needToUpdateShadow = False
        needToWriteToBelt = False

        if PROXIMITY_STRING + "{" in data:
            beltData = json.loads(data.split(PROXIMITY_STRING)[1])

            if "mode" in beltData and self.SHADOW_REPORTED["mode"] != beltData["mode"]:
                self.SHADOW_REPORTED["mode"] = beltData["mode"]
                needToUpdateShadow = True

            if "speed" in beltData and self.SHADOW_REPORTED["speed"] != beltData["speed"]:
                self.SHADOW_REPORTED["speed"] = beltData["speed"]
                needToUpdateShadow = True

            if self.SENSORS["proximity"]["sensor1"] != beltData["sensor1"] or self.SENSORS["proximity"]["sensor2"] != beltData["sensor2"]:
                self.SENSORS["proximity"]["sensor1"] = beltData["sensor1"]
                self.SENSORS["proximity"]["sensor2"] = beltData["sensor2"]

                GGIOT.publish(topic="{}/proximity".format(TOPIC_FOR_SENSORS), payload={
                    "sensor1": self.SENSORS["proximity"]["sensor1"],
                    "sensor2": self.SENSORS["proximity"]["sensor2"],
                    "mode": self.SHADOW_REPORTED["mode"],
                    "speed": self.SHADOW_REPORTED["speed"]
                })

        elif TELEMETRY_STRING + "{" in data:
            beltData = json.loads(data.split(TELEMETRY_STRING)[1])

            if "chassis" in beltData and (self.SENSORS["chassis"]["x"] != beltData["chassis"]["x"] or self.SENSORS["chassis"]["y"] != beltData["chassis"]["y"] or self.SENSORS["chassis"]["z"] != beltData["chassis"]["z"]):
                self.SENSORS["chassis"] = beltData["chassis"]
                GGIOT.publish(topic="{}/chassis".format(TOPIC_FOR_SENSORS), payload=beltData["chassis"])

            if "speed" in beltData and self.SENSORS["speed"]["rpm"] != beltData["speed"]["rpm"]:
                self.SENSORS["speed"]["rpm"] = beltData["speed"]["rpm"]
                GGIOT.publish(topic="{}/speed".format(TOPIC_FOR_SENSORS), payload=beltData["speed"])

        elif SHADOW_STRING + "{" in data:
            beltData = json.loads(data.split(SHADOW_STRING)[1])

            if "state" in beltData and "reported" in beltData["state"]:
                reported = beltData["state"]["reported"]
                # print("Belt.parseSerial: reported: {}".format(json.dumps(reported)))

                if "speed" in reported:
                    if reported["speed"] != 1 and reported["speed"] != 2:
                        needToWriteToBelt = True
                    elif self.SHADOW_REPORTED["speed"] != reported["speed"]:
                        self.SHADOW_REPORTED["speed"] = reported["speed"]
                        needToUpdateShadow = True

                if "mode" in reported:
                    if reported["mode"] != 1 and reported["mode"] != 2 and reported["mode"] != 3:
                        needToWriteToBelt = True
                    elif self.SHADOW_REPORTED["mode"] != reported["mode"]:
                        self.SHADOW_REPORTED["mode"] = reported["mode"]
                        needToUpdateShadow = True

                if needToWriteToBelt or self.SHADOW_REPORTED["mode"] != self.SHADOW_DESIRED["mode"] or self.SHADOW_REPORTED["speed"] != self.SHADOW_DESIRED["speed"]:
                    self.writeDesiredToSerial(serialWrite)

        if needToUpdateShadow == True:
            GGIOT.updateThingShadow(payload={"state": {"reported": self.SHADOW_REPORTED}})


    def parseIncomingShadow(self, data):

        needToWriteToBelt = False

        if data and "state" in data and "desired" in data["state"]:

            desired = data["state"]["desired"]

            print("Belt.parseIncomingShadow: desired: {}".format(json.dumps(desired)))

            if "mode" in desired:
                self.SHADOW_DESIRED["mode"] = int(desired["mode"])
                needToWriteToBelt = True

            if "speed" in desired:
                self.SHADOW_DESIRED["speed"] = int(desired["speed"])
                needToWriteToBelt = True

        return needToWriteToBelt
